Composer/packages/client/src/components/TriggerCreationModal/TriggerDropdownGroup.tsx

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+/** @jsx jsx */
+import { jsx } from '@emotion/core';
+import { FC, ReactNode, useCallback, useMemo, useState } from 'react';
+import { Stack } from 'office-ui-fabric-react/lib/Stack';
+import { Dropdown } from 'office-ui-fabric-react/lib/Dropdown';
+import { IDropdownOption } from 'office-ui-fabric-react/lib/Dropdown';
+import { Icon } from 'office-ui-fabric-react/lib/Icon';
+import { SDKKinds } from '@bfc/shared';
+import { useTriggerConfig } from '@bfc/extension-client';
+import formatMessage from 'format-message';
+
+import { dropdownStyles, optionStyles, warningIconStyles } from './styles';
+import {
+  generateTriggerOptionTree,
+  TriggerOptionGroupNode,
+  TriggerOptionLeafNode,
+  TriggerOptionTreeNode,
+} from './TriggerOptionTree';
+import { checkRecognizerCompatibility } from './checkRecognizerCompatibility';
+import { triggerOrderMap } from './schema/defaultTriggerOrder';
+
+export interface TriggerDropwdownGroupProps {
+  recognizerType: SDKKinds | undefined;
+  triggerType: string;
+  setTriggerType: (type: string) => void;
+}
+
+const getTriggerLabelOrder = (label: string) => {
+  return triggerOrderMap[label] ?? Number.MAX_VALUE;
+};
+
+export const TriggerDropdownGroup: FC<TriggerDropwdownGroupProps> = ({ recognizerType, setTriggerType }) => {
+  const renderDropdownOption = useCallback(
+    (option?: IDropdownOption) => {
+      if (!option) return null;
+      const compatible = checkRecognizerCompatibility(option.key as SDKKinds, recognizerType);
+      return (
+        <div css={optionStyles}>
+          {option.text}
+          {!compatible && <Icon iconName={'warning'} style={warningIconStyles} />}
+        </div>
+      );
+    },
+    [recognizerType]
+  );
+
+  const triggerUISchema = useTriggerConfig();
+  const triggerOptionTree = useMemo(() => {
+    return generateTriggerOptionTree(
+      triggerUISchema,
+      formatMessage('What is the type of this trigger?'),
+      formatMessage('Select a trigger type'),
+      (opt1, opt2) => getTriggerLabelOrder(opt1.label) - getTriggerLabelOrder(opt2.label)
+    );
+  }, []);
+
+  const [activeNode, setActiveNode] = useState<TriggerOptionTreeNode>(triggerOptionTree);
+  const onClickNode = (node: TriggerOptionTreeNode) => {
+    setActiveNode(node);
+    if (node instanceof TriggerOptionLeafNode) {
+      setTriggerType(node.$kind);
+    } else {
+      setTriggerType('');
+    }
+  };
+
+  const getDropdownList = (activeNode: TriggerOptionTreeNode) => {
+    const treePath: TriggerOptionTreeNode[] = [activeNode];
+    while (treePath[0].parent) {
+      treePath.unshift(treePath[0].parent);
+    }
+
+    const dropdownList: ReactNode[] = [];
+
+    const getKey = (x: TriggerOptionTreeNode) => (x instanceof TriggerOptionLeafNode ? x.$kind : x.label);
+
+    for (let i = 0; i < treePath.length; i++) {
```

Comment:
use `map` + `filter` or `reduce` instead.

Reply:
This loop needs to look up current item and the next item to build a Dropdown, so I made it a indexed based loop